x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/SingleUseKey.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.repositories.encrypted;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.collect.Tuple;
+
+import javax.crypto.SecretKey;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Container class for a {@code SecretKey} with a unique identifier, and a 4-byte wide {@code Integer} nonce, that can be used for a
+ * single encryption operation. Use {@link #createSingleUseKeySupplier(CheckedSupplier)} to obtain a {@code Supplier} that returns
+ * a new {@link SingleUseKey} instance on every invocation. The number of unique {@code SecretKey}s (and their associated identifiers)
+ * generated is minimized and, at the same time, ensuring that a given {@code nonce} is not reused with the same key.
+ */
+final class SingleUseKey {
+    private static final Logger logger = LogManager.getLogger(SingleUseKey.class);
+    static final int MIN_NONCE = Integer.MIN_VALUE;
+    static final int MAX_NONCE = Integer.MAX_VALUE;
+    private static final int MAX_ATTEMPTS = 9;
+    private static final SingleUseKey EXPIRED_KEY = new SingleUseKey(null, null, MAX_NONCE);
+
+    private final BytesReference keyId;
+    private final SecretKey key;
+    private final int nonce;
+
+    // for tests use only!
+    SingleUseKey(BytesReference KeyId, SecretKey Key, int nonce) {
+        this.keyId = KeyId;
+        this.key = Key;
+        this.nonce = nonce;
+    }
+
+    public BytesReference getKeyId() {
+        return keyId;
+    }
+
+    public SecretKey getKey() {
+        return key;
+    }
+
+    public int getNonce() {
+        return nonce;
+    }
+
+    /**
+     * Returns a {@code Supplier} of {@code SingleUseKey}s so that no two instances contain the same key and nonce pair.
+     * A new key is generated only when the {@code nonce} space has been exhausted.
+     */
+    static <T extends Exception> CheckedSupplier<SingleUseKey, T> createSingleUseKeySupplier(
+        CheckedSupplier<Tuple<BytesReference, SecretKey>, T> keyGenerator
+    ) {
+        final AtomicReference<SingleUseKey> keyCurrentlyInUse = new AtomicReference<>(EXPIRED_KEY);
+        return createSingleUseKeySupplier(keyGenerator, keyCurrentlyInUse);
+    }
+
+    // for tests use only
+    static <T extends Exception> CheckedSupplier<SingleUseKey, T> createSingleUseKeySupplier(
+        CheckedSupplier<Tuple<BytesReference, SecretKey>, T> keyGenerator,
+        AtomicReference<SingleUseKey> keyCurrentlyInUse
+    ) {
+        final Object lock = new Object();
+        return () -> {
```

Comment:
This feels like a complicated way to produce a `Supplier` of a set of keys with a sequential nonce.
Does it really need the complexity of `AtomicReference` over a `synchronized` method? 

It feels like the general concept could be written quite simply with a class that extends `Supplier`, has an internal counter and is synchronized on `get` (or uses an `AtomicInteger` for the counter).
I presume there is a reason why you took this approach instead, but the code doesn't tell me why that is.

Reply:
> It feels like the general concept could be written quite simply with a class that extends Supplier, has an internal counter and is synchronized on get (or uses an AtomicInteger for the counter).

This is pretty much what the code here does; here it is without production ornaments:
```
        return () -> {
                final SingleUseKey nonceAndKey = keyCurrentlyInUse.getAndUpdate(
                    prev -> prev.nonce < MAX_NONCE ? new SingleUseKey(prev.keyId, prev.key, prev.nonce + 1) : EXPIRED_KEY
                );
                if (nonceAndKey.nonce < MAX_NONCE) {
                    return nonceAndKey;
                } else {
                    synchronized (lock) {
                        if (keyCurrentlyInUse.get().nonce == MAX_NONCE) {
                            final Tuple<BytesReference, SecretKey> newKey = keyGenerator.get();
                            keyCurrentlyInUse.set(new SingleUseKey(newKey.v1(), newKey.v2(), MIN_NONCE));
                        }
                    }
                }
        };
```

Instead of extending Supplier this uses the fact that Supplier is a `FunctionalInterface`. The internal counter is the nonce itself. Instead of synchronizing on get this uses an `AtomicReference` as an `AtomicInteger`. The remaining code is to generate a new key and reset the counter. And in this last code is where the preference for `AtomicReference` over of `AtomicInteger` shows off because it allows to atomically set the key and the nonce.

I have added a few more comments. I hope this all makes more sense now.
Do you think I should change something here?

Reply:
It makes sense. I think I was missing the fact that you need to generate a new base key when the nonce rolls over. but I do think this code is unnecessarily difficult to follow.
That's a combination of trying too hard to do it all in a lambda and trying to keep the locks really small. Is that necessary?

Couldn't we have:
```
        return new CheckedSupplier<SingleUseKey, T>() {
                private Tuple<BytesReference, SecretKey> key = null;
                private int nonce = 0;
                public synchronized SingleUseKey get() throws T() {
                   if( key == null || nonce == MAX_NONCE) {
                       key = keyGenerator.get();
                       nonce = MIN_NONCE;
                   } else {
                       nonce++;
                   }
                   return new SingleUseKey(key.v1(), key.v2(), nonce);
                }
        };
```